webpush/encrypt.go

```diff
@@ -270,7 +281,7 @@ func hkdf(salt, ikm, info []byte, length int) []byte {
 	mac = hmac.New(sha256.New, prk)
 	mac.Write(info)
 	mac.Write([]byte{1})
-	return mac.Sum(nil)[0:length]
+	return mac.Sum(nil)[0:length], nil
```

Comment:
What this function does is create a SHA256 hash of some data and then return the first `length` bytes of it. Because a SHA256 hash is 32 bytes long, a requested `length` of > 32 will fail.

The actual lengths used by web push are [32](https://github.com/GoogleChrome/push-encryption-go/blob/master/webpush/encrypt.go#L177), [16](https://github.com/GoogleChrome/push-encryption-go/blob/master/webpush/encrypt.go#L196) and [12](https://github.com/GoogleChrome/push-encryption-go/blob/master/webpush/encrypt.go#L201).

If a future version of the spec requires some value to be created with an HKDF length of more than 32, I really think that it is better to fail obviously (big error, length was too big) than fail subtly (everything seemed to work but the client rejected the push message). The internal security audit of this code had this as a suggestion, which is where the issue came from.

However, I also want to get the rest merged so if I haven't convinced you yet then I'll do whatever you think best from here :)